core/src/main/java/io/grpc/internal/GzipInflatingBuffer.java

```diff
@@ -300,215 +287,162 @@ private int fillInflatedBuf(int n) throws DataFormatException, ZipException {
   }
 
-  private boolean processTrailer() throws ZipException {
-    if (gzipMetadataReader.readableBytes() >= GZIP_TRAILER_SIZE) {
-      if (crc.getValue() != (readUnsignedInt(gzipMetadataReader))
-          ||
-          (readUnsignedInt(gzipMetadataReader) != expectedGzipTrailerIsize)) {
-        throw new ZipException("Corrupt GZIP trailer");
-      }
-      crc.reset();
-      state = State.HEADER_MAGIC;
-      return true;
+  private boolean initializeInflater() {
+    if (inflater == null) {
+      inflater = new Inflater(true);
     } else {
-      return false;
+      inflater.reset();
     }
+    crc.reset();
+    int bytesRemainingInInflaterInput = inflaterInputEnd - inflaterInputStart;
+    if (bytesRemainingInInflaterInput > 0) {
+      inflater.setInput(inflaterInput, inflaterInputStart, bytesRemainingInInflaterInput);
+      state = State.INFLATING;
+    } else {
+      state = State.INFLATER_NEEDS_INPUT;
+    }
+    return true;
   }
 
-  /** Skip over a zero-terminated byte sequence. Returns true when the zero byte is read. */
-  private boolean readBytesUntilZero() {
-    while (gzipMetadataReader.readableBytes() > 0) {
-      if (gzipMetadataReader.readUnsignedByte() == 0) {
-        return true;
+  private int inflate(byte[] b, int off, int len) throws DataFormatException, ZipException {
+    checkState(inflater != null, "inflater is null");
+
+    try {
+      int inflaterTotalIn = inflater.getTotalIn();
+      int n = inflater.inflate(b, off, len);
+      int bytesConsumedDelta = inflater.getTotalIn() - inflaterTotalIn;
+      bytesConsumed += bytesConsumedDelta;
+      inflaterInputStart += bytesConsumedDelta;
+      crc.update(b, off, n);
+
+      if (inflater.finished()) {
+        // Save bytes written to check against the trailer ISIZE
+        expectedGzipTrailerIsize = (inflater.getBytesWritten() & 0xffffffffL);
+        if (gzipMetadataReader.readableBytes() <= GZIP_HEADER_MIN_SIZE + GZIP_TRAILER_SIZE) {
+          // We don't have enough bytes to begin inflating a concatenated gzip stream, drop context
+          inflater.end();
+          inflater = null;
+        }
+        state = State.TRAILER;
+      } else if (inflater.needsInput()) {
+        state = State.INFLATER_NEEDS_INPUT;
       }
+
+      return n;
+    } catch (DataFormatException e) {
+      // Wrap the exception so tests can check for a specific prefix
+      throw new DataFormatException("Inflater data format exception: " + e.getMessage());
     }
-    return false;
   }
 
-  /** Reads unsigned short in Little-Endian byte order. */
-  private int readUnsignedShort(GzipMetadataReader buffer) {
-    return buffer.readUnsignedByte() | (buffer.readUnsignedByte() << 8);
+  private boolean fill() {
+    checkState(inflater != null, "inflater is null");
+    checkState(inflaterInputStart == inflaterInputEnd, "inflaterInput has unconsumed bytes");
+    int bytesToAdd = Math.min(gzippedData.readableBytes(), INFLATE_BUFFER_SIZE);
+    if (bytesToAdd > 0) {
```

Comment:
invert condition and return `false` early to avoid the indentation in the successful case.

Reply:
Done